---
title: "Differential gene expression"
author: "Vanessa"
date: '2024-03-22'
output:
  word_document: default
  pdf_document:
    latex_engine: xelatex
  html_document: default
---

## \*\* Introduction \*\*

In this code, we will use EDGER to analyze differentially expressed genes in RNA-Seq samples from the arcuate nucleus of the hypothalamus of exercised mice.

### Groups:

CTL, 0h, 2h and 4h

------------------------------------------------------------------------

## **1.0 Loadings**

``` r
library(DESeq2)
library(ggplot2)
library(edgeR)
library(factoextra)
library(devtools)
library(tidyverse)
library(dplyr)
library(knitr)
library(stringr)
library(pheatmap)
library(RColorBrewer)
library(ggsci)
library(biomaRt)
library(dendextend)
library(tibble)
```

------------------------------------------------------------------------

### **1.2 Loading Data**

Loading:

1.  `eDat` - Count matrix: quantification_raw
2.  `gDat` - Group file

``` r
eDat <- read.table("Count.txt", header=TRUE, sep="\t", row.names = 1)
gDat <- read.table("Group.txt", header=TRUE, sep="\t")


# exploring data

# View data
head(eDat)
knitr::kable(head(eDat))
```

| CTL_1         | CTL_2 | CTL_3 | CTL_4 | CTL_5 | Exe0h_1 | Exe0h_2 | Exe0h_3 | Exe0h_4 | Exe0h_5 | Exe2h_1 | Exe2h_2 | Exe2h_3 | Exe2h_4 | Exe2h_5 | Exe4h_1 | Exe4h_2 | Exe4h_3 | Exe4h_4 | Exe4h_5 |     |
|:---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 0610005C13Rik |     5 |     1 |     8 |     5 |       8 |       4 |       9 |       6 |      10 |       3 |       3 |       5 |       9 |       3 |      11 |       6 |       3 |       3 |       4 |   7 |
| 0610006L08Rik |     0 |     0 |     0 |     0 |       0 |       0 |       0 |       0 |       0 |       0 |       0 |       0 |       0 |       0 |       0 |       1 |       0 |       0 |       0 |   0 |
| 0610009E02Rik |    21 |     8 |    24 |     3 |      13 |      12 |      13 |      16 |      15 |      13 |       8 |      17 |      12 |       5 |      16 |      16 |       7 |      17 |      10 |  12 |
| 0610009L18Rik |    44 |    32 |    32 |    36 |      28 |      24 |      51 |      42 |      61 |      42 |      17 |      40 |      51 |      42 |      54 |      50 |      32 |      21 |      27 |  30 |
| 0610010K14Rik |   401 |   221 |   327 |   230 |     220 |     218 |     423 |     348 |     482 |     472 |     211 |     398 |     453 |     325 |     438 |     451 |     227 |     212 |     215 | 388 |
| 0610025J13Rik |     2 |     1 |     1 |     1 |       2 |       3 |       3 |       0 |       2 |       2 |       0 |       1 |       0 |       1 |       1 |       2 |       1 |       0 |       2 |   1 |

``` r
# View data
head(gDat)
knitr::kable(head(gDat))
```

| sample  | treat |
|:--------|:------|
| CTL_1   | CTL   |
| CTL_2   | CTL   |
| CTL_3   | CTL   |
| CTL_4   | CTL   |
| CTL_5   | CTL   |
| Exe0h_1 | Exe0h |

### **1.3 Exploring dataset**

``` r
## How many transcripts are there in the entire library?

dim(eDat)
```

[1] 41079 20

``` r
## Count number of reads in each library
knitr::kable(colSums(eDat))  
```

|         |        x |
|:--------|---------:|
| CTL_1   | 19659118 |
| CTL_2   | 12709869 |
| CTL_3   | 16504944 |
| CTL_4   | 10506117 |
| CTL_5   | 12207032 |
| Exe0h_1 | 10999794 |
| Exe0h_2 | 23318452 |
| Exe0h_3 | 17763551 |
| Exe0h_4 | 22971441 |
| Exe0h_5 | 22150547 |
| Exe2h_1 |  9637053 |
| Exe2h_2 | 18571904 |
| Exe2h_3 | 22277769 |
| Exe2h_4 | 18034142 |
| Exe2h_5 | 21239057 |
| Exe4h_1 | 19644975 |
| Exe4h_2 | 11842915 |
| Exe4h_3 | 12979164 |
| Exe4h_4 | 13180388 |
| Exe4h_5 | 15398202 |

```{r dend-plot, fig.cap=""}
## Load necessary libraries
library(ggdendro)
library(ggplot2)

# Function to plot dendrogram with legend
plot_dendrogram <- function(d, title, legend_labels, legend_colors) {
  plot(d, main = title)
  
  # Get plot coordinates
  coords <- par("usr")
  
  # Set legend coordinates
  legend_x <- coords[2] - 0.055 * diff(coords[1:2])
  legend_y <- coords[4] + 0.05 * diff(coords[3:4])
  
  # Add legend outside the plot
  legend(legend_x, legend_y, legend = legend_labels, fill = legend_colors, bty = "n", xpd = TRUE)
}

# Load data
eDat <- read.table("Count.txt", header=TRUE, sep="\t", row.names = 1)

# Calculate Spearman correlation
spear <- cor(eDat, method="spearman")

# Calculate distance matrix
d <- as.dist(1 - spear)

# Perform hierarchical clustering
upgma <- hclust(d, method="average")

# Function to set label color
labelCol <- function(x) {
  if (is.leaf(x)) {
    label <- attr(x, "label")
    color <- ifelse(grepl("^CTL", label), "red", 
                    ifelse(grepl("^Exe0h", label), "blue",
                           ifelse(grepl("^Exe2h", label), "green", "purple")))
    attr(x, "nodePar") <- list(lab.col=color)
  }
  return(x)
}

# Apply labelCol on all nodes of the dendrogram
d <- dendrapply(as.dendrogram(upgma), labelCol)

## Plot dendrogram
plot_dendrogram(d, "Dendrogram before normalization", c("CTL", "0h", "2h", "4h"), c("red", "blue", "green", "purple"))

# Save plot
png("plot1.png", width=7, height=7, units="in", res=300)
plot_dendrogram(d, "Dendrogram before normalization", c("CTL", "0h", "2h", "4h"), c("red", "blue", "green", "purple"))
dev.off()


## Load necessary libraries
library(limma)
library(edgeR)

# Increase right margin
par(mar = c(5, 4, 4, 6) + 0.1)

# Define colors to the groups
colors <- c(rep("red", 5), rep("blue", 5), rep("green", 5), rep("purple", 5))

## Plot MDS plot
plotMDS(eDat, col = colors, pch = 16)

# Get the plot coordinates
plot_coord <- par("usr")

# Define coordinates for the legend
x_legend <- plot_coord[2] + 0.01 * diff(plot_coord[2:1])  # 1% to the left of the plot
y_legend <- plot_coord[4] - 0.05 * diff(plot_coord[4:3])  # 5% above the plot

# Add legend outside the plot
legend(x = x_legend, y = y_legend, legend = c("CTL", "0h", "2h", "4h"), 
       col = c("red", "blue", "green", "purple"), pch = 16, title = "", bty = 'n', 
       xpd = TRUE)


# Save plot
# Set up PNG device with desired resolution
png("plot2.png", width = 800, height = 600, units = "px", res = 300)

# Increase right margin
par(mar = c(5, 4, 4, 6) + 0.1)

# Define colors to the groups
colors <- c(rep("red", 5), rep("blue", 5), rep("green", 5), rep("purple", 5))

# Perform MDS plot
plotMDS(eDat, col = colors, pch = 16)

# Get the plot coordinates
plot_coord <- par("usr")

# Define coordinates for the legend
x_legend <- plot_coord[2] + 0.01 * diff(plot_coord[2:1])  # 1% to the left of the plot
y_legend <- plot_coord[4] - 0.05 * diff(plot_coord[4:3])  # 5% above the plot

# Add legend outside the plot
legend(x = x_legend, y = y_legend, legend = c("CTL", "0h", "2h", "4h"), 
       col = c("red", "blue", "green", "purple"), pch = 16, title = "", bty = 'n', 
       xpd = TRUE)

# Save the plot
dev.off()


## Boxplot before normalization (outliers)

# Increase right margin
par(mar = c(5, 6, 4, 4) + 0.1)

# Get the plot coordinates
plot_coord <- par("usr")

# plot
boxplot(eDat, col = colors, outline = T, pch = 20,
        main="Read counts number per gene for each sample\n (outliers)",
        xlab="",
        ylab="", las=2,cex.axis=0.8)

# Move y-axis label further to the left
mtext(side = 2, text = "Read counts number per gene", line = 4)


## Box plot without outliers
# plot
boxplot(eDat, col = colors, outline = F, pch = 20,
        main="Read counts number per gene for each sample\n (without outliers)",
        xlab="",
        ylab="", las=2,cex.axis=0.8)

# Move y-axis label further to the left
mtext(side = 2, text = "Read counts number per gene", line = 4)


# Get plot coordinates
coords <- par("usr")

# Increase right margin
par(mar = c(5, 6, 4, 7) + 0.1)

# Set legend coordinates
legend_x <- coords[2] - 0.001 * diff(coords[1:2])
legend_y <- coords[4] + 0.05 * diff(coords[3:4])

# Add legend outside the plot
legend(legend_x, legend_y, legend = c("CTL", "0h", "2h", "4h"), 
       col = c("red", "blue", "green", "purple"), pch = 16, title = "", bty = 'n', 
       xpd = TRUE)



#Pheatmpa Distances
library(pheatmap)
d_matrix <- as.matrix(eDat)
pheatmap(d_matrix, show_rownames=F, show_colnames=T)

```

------------------------------------------------------------------------

### **1.4 Groups (CTL vs 0 hours)**

```{r plot, fig.cap=""}
library(tibble)
library(ggrepel)

eDat <- read.table("Count.txt", header=TRUE, sep="\t", row.names = 1)
gDat <- read.table("Group.txt", header=TRUE, sep="\t")


## Counts
## filter only samples CTL and Exe0h
suppressMessages({library(dplyr)})

# Select only 'CTL' ou 'Exe0h'
selected_columns <- eDat[, grepl("^CTL|^Exe0h", names(eDat))]
  
# new table
new_table1 <- cbind(rownames = rownames(eDat), selected_columns)

## remove the first colllumn
new_table1$rownames <- NULL

## Treat
new_group1 <- gDat[1:10, ]

## load EdgeR ##
library(edgeR)

# ATENTION!!
new_group1$treat

# Control treatment needs to appear as the first level for the Treat Vs. Control comparison
new_group1$treat <-factor(new_group1$treat, levels=c("CTL", "Exe0h"))

# Input data
y_adj <- DGEList(counts=new_table1, samples=new_group1, group=new_group1$treat)
design_adj <- model.matrix(~ treat - 1, data = new_group1)

# Filter lowly expressed genes
keep_adj <- rowSums(cpm(y_adj)>1) >= 2
sum(keep_adj)
y_adj <- y_adj[keep_adj, , keep.lib.sizes=FALSE]

## Explanation: ">= 2" is a condition to maintain the gene, which means that the gene needs to have at least two replicas with CPM greater than 1 to be maintained.

## Explanation: after filtering, 16131 genes remained that have sufficient expression in at least two replicates with CPM greater than 1. Genes that do not meet these criteria are removed from the dataset.


# Normalization for RNA composition
y_adj <- calcNormFactors(y_adj)
y_adj$samples
norm.expr_adj <- y_adj$samples

# norm counts
logcpm_adj <- cpm(y_adj, log=TRUE)


##########################################
##      Data analyses with edgeR        ##
##########################################


# Estimate dispersion with Cox-Reid profile-adjusted likelihood (CR) method
y_adj <- estimateDisp(y_adj, design_adj)
y_adj$common.dispersion


## Explanation: A low dispersion value indicates that the expression levels of genes are relatively consistent across replicates, suggesting that there is little variability beyond what is expected due to random sampling. Conversely, a high dispersion value indicates greater variability in gene expression levels among replicates.


## plot
plotBCV(y_adj)

# Fit Generalized Linear Model (GLM) 
fit <- glmFit(y_adj, design_adj)
colnames(fit)

# Differential gene expression analysis
lrt <- glmLRT(fit, contrast=c(-1,1))

## Coefficients used in the contrast for the likelihood ratio test
topTags(lrt)

## Explanation: It indicates that the comparison is between the coefficient for treatCTL (the reference group) and the coefficient for treatExe0h (the group of interest).

## Number of genes identified as differentially expressed (Disregarding parameters)
is.de <- decideTestsDGE(lrt)
summary(decideTestsDGE(lrt))

## plot MD
plotMD(lrt, status=is.de)

# Number of up and down-regulated. FDR < 5% and |log2 fold change| ≥ 1
summary(de <- decideTestsDGE(lrt, lfc=0))

## Explanation: The absolute value of 1 on this scale indicates that there was a doubling (2-fold) change in gene expression between the two conditions.

# MA-plot
detags <- rownames(y_adj)[as.logical(de)]
plotSmear(lrt, de.tags=detags)
abline(h=c(-0, 0), col="blue")

# To export the results
result <- as.data.frame(topTags(lrt, n=1000000))
result <- add_column(result, symbol= rownames(result), .before = "logFC")
     
## save as table
write.table(result, "DEGs_CTL_vs_0h.txt", sep = "\t", row.names = FALSE)


## Check all the genes Downregulated and Upregulated

# Remove unnecessary columns
result$logCPM <- NULL
result$LR <- NULL

# Remove rows with FDR < 0.05 and logFC >= 0
result$expression <- ifelse(result$FDR < 0.05 & abs(result$logFC) >= 0,
                        ifelse(result$logFC >= 0, 'Up', 'Down'),
                        'Stable')

# Select only genes that are upregulated or downregulated
result_final <- subset(result, expression %in% c("Up", "Down"))

# Remove row names
rownames(result_final) <- NULL

## View the table
knitr::kable(result_final)  


##########################################
############      PLOTS        ###########
##########################################

## volcano plot

# Extract top tags from lrt
result_edgeR <- as.data.frame(topTags(lrt, n=1000000, adjust.method = "BH"))

# Extract log-transformed counts per million
count_CPM <- as.data.frame(logcpm_adj)

## Convert row names into first column
result_edgeR <- add_column(result_edgeR, gene = rownames(result_edgeR), .before = "logFC")

count_CPM <- add_column(count_CPM, gene = rownames(count_CPM), .before = "CTL_1")

## Order by gene
result_edgeR <- result_edgeR[order(result_edgeR$gene),]
count_CPM <- count_CPM[order(count_CPM$gene),]

## Combine result_edgeR columns with count_CPM
count_CPM <- add_column(count_CPM, logFC = result_edgeR$logFC, .before = "CTL_1")
count_CPM <- add_column(count_CPM, FDR = result_edgeR$FDR, .before = "CTL_1")
count_CPM <- add_column(count_CPM, gene2 = result_edgeR$gene, .before = "CTL_1")

# Check if gene and gene2 are the same and in the same order
if(all(count_CPM$gene == count_CPM$gene2)) {
  print("The values in 'gene' and 'gene2' are the same.")
} else {
  print("The values in 'gene' and 'gene2' are different.")
  different_indices <- which(count_CPM$gene != count_CPM$gene2)
  print(paste("The value in 'gene' is different at indices:", different_indices))
}

## remove gene2
count_CPM$gene2 <- NULL

## Insert -log10 FDR
count_CPM <- add_column(count_CPM, logFDR = -log10(result_edgeR$FDR), .before = "CTL_1")

## Insert -log10 p-value
count_CPM <- add_column(count_CPM, logpvalue = -log10(result_edgeR$PValue), .before = "CTL_1")


## PLOT

# Assign count_CPM to expressao_genica
expressao_genica <- count_CPM

# Determine expression based on FDR and logFC thresholds
expressao_genica$expression <- ifelse(expressao_genica$FDR < 0.05 & abs(expressao_genica$logFC) >= 0,
                                      ifelse(expressao_genica$logFC >= 0, 'Up', 'Down'),
                                      'Stable')

# Count the number of genes categorized as Up and Down
sum(expressao_genica$expression == "Up")
sum(expressao_genica$expression == "Down")

# Create volcano plot

# Filter the data to include only "Down" and "Up" genes
expressao_genica_filtered <- expressao_genica[expressao_genica$expression %in% c("Down", "Up"), ]

# Order the filtered dataframe by logFC
expressao_genica_filtered <- expressao_genica_filtered[order(expressao_genica_filtered$logFC, decreasing = TRUE), ]

# Select the top 5 upregulated genes and the top 5 downregulated genes
top_genes_up <- head(subset(expressao_genica_filtered, expression == "Up"), 5)
top_genes_down <- head(subset(expressao_genica_filtered, expression == "Down"), 5)

# Combine top_genes_up and top_genes_down into a single dataframe
top_genes <- rbind(top_genes_up, top_genes_down)

## see
print(top_genes)

# volcano plot
plot1 <- ggplot(data = expressao_genica,
       aes(x = logFC,
           y = logFDR,
           colour = expression)) +
  geom_point(size = 3) +
  geom_label_repel(data = top_genes, aes(label = gene), size = 5, force = 5, 
                   fill = "white", color = "black", box.padding = unit(0.5, "lines")) +  # Labeling top 10 genes with a rectangle
  scale_color_manual(values = c("red", "grey", "#009933"),
                     name = "",
                     breaks = c("Up", "Stable", "Down")) +
  geom_vline(xintercept = c(-0, 0), lty = 4, col = "black", lwd = 0.8) +
  labs(x = "log2FC",
       y = "-log10 (adj.P.Val)") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "right",
        legend.title = element_blank(),
        text = element_text(size = 12)) +
  ggtitle("CTL vs 0 hours")

# To display the volcano plot
plot1

## save as image
tiff("volcano_CTL_vs_0h.tiff", units="in", width=7, height=7, res=300)
plot1 
dev.off()


## heatmap

## Only genes Down and Upregulated
head(expressao_genica_filtered)

#remove the fisrt five cols
expressao_genica_filtered <- expressao_genica_filtered[, -(1:5)]

## remove a col name expression
expressao_genica_filtered$expression <- NULL

# Set column names
colnames(expressao_genica_filtered) <- c(rep("CTL", 5), rep("0h", 5)) 

# Convert dataframe to matrix
samp.with.rownames <- as.matrix(expressao_genica_filtered)

# Create column annotations for heatmap
suppressWarnings({
column_annotation <- c(rep("#74ADD1", 5), rep("#A6D96A", 5))
column_annotation <- as.matrix(column_annotation)
colnames(column_annotation) <- c("Variable X")
})

# plot
library("gplots")

heatmap.2(samp.with.rownames, scale="row", trace = "none", cexRow=1.5, cexCol=2, margins = c(5, 11), colSideColors=samp.with.rownames,ColSideColors=column_annotation, labCol = FALSE)

# Get plot coordinates
coords <- par("usr")

# Increase right margin
par(mar = c(5, 6, 4, 12) + 0.2)

# Set legend coordinates
legend_x <- coords[2] + 0.2 * diff(coords[1:2])
legend_y <- coords[4] + 0.05 * diff(coords[3:4])

# Add legend outside the plot
legend(legend_x, legend_y, legend = c("CTL", "0h"), 
       fill = c("#74ADD1", "#A6D96A"), border = FALSE, cex = 1.1, 
       xpd = TRUE)


## save as image
suppressWarnings({
tiff("heatmap_CTL_vs_0h.tiff", units="in", width=7, height=7, res=300)
heatmap.2(samp.with.rownames, scale="row", trace = "none", cexRow=1.5, cexCol=2, margins = c(5, 11), colSideColors=samp.with.rownames,ColSideColors=column_annotation, labCol = FALSE)

# Get plot coordinates
coords <- par("usr")

# Increase right margin
par(mar = c(5, 6, 4, 12) + 0.2)

# Set legend coordinates
legend_x <- coords[2] + 0.2 * diff(coords[1:2])
legend_y <- coords[4] + 0.05 * diff(coords[3:4])

# Add legend outside the plot
legend(legend_x, legend_y, legend = c("CTL", "0h"), 
       fill = c("#74ADD1", "#A6D96A"), border = FALSE, cex = 1.1, 
       xpd = TRUE)
dev.off()
})


```

------------------------------------------------------------------------


### **1.5 Groups (CTL vs 2 hours)**


```{r plot2, fig.cap=""}
suppressWarnings({
  library(tibble)
  library(dplyr)
  library(edgeR)
  library(ggrepel)
  })

## Import dataset
eDat <- read.table("Count.txt", header=TRUE, sep="\t", row.names = 1)
gDat <- read.table("Group.txt", header=TRUE, sep="\t")


## Counts
## filter only samples CTL and Exe2h

# Select only 'CTL' ou 'Exe2h'
selected_columns <- eDat[, grepl("^CTL|^Exe2h", names(eDat))]
  
# new table
new_table1 <- cbind(rownames = rownames(eDat), selected_columns)

## remove the first colllumn
new_table1$rownames <- NULL

## Treat
new_group1 <- subset(gDat, treat == "CTL" | treat == "Exe2h")

# ATENTION!!
new_group1$treat

# Control treatment needs to appear as the first level for the Treat Vs. Control comparison
new_group1$treat <-factor(new_group1$treat, levels=c("CTL", "Exe2h"))

# Input data
y_adj <- DGEList(counts=new_table1, samples=new_group1, group=new_group1$treat)
design_adj <- model.matrix(~ treat - 1, data = new_group1)

# Filter lowly expressed genes
keep_adj <- rowSums(cpm(y_adj)>1) >= 2
sum(keep_adj)
y_adj <- y_adj[keep_adj, , keep.lib.sizes=FALSE]


## Explanation: ">= 2" is a condition to maintain the gene, which means that the gene needs to have at least two replicas with CPM greater than 1 to be maintained.

## Explanation: after filtering, 16010 genes remained that have sufficient expression in at least two replicates with CPM greater than 1. Genes that do not meet these criteria are removed from the dataset.

# Normalization for RNA composition
y_adj <- calcNormFactors(y_adj)
y_adj$samples
norm.expr_adj <- y_adj$samples

# norm counts
logcpm_adj <- cpm(y_adj, log=TRUE)

##########################################
##      Data analyses with edgeR        ##
##########################################


# Estimate dispersion with Cox-Reid profile-adjusted likelihood (CR) method
y_adj <- estimateDisp(y_adj, design_adj)
y_adj$common.dispersion


## Explanation: A low dispersion value indicates that the expression levels of genes are relatively consistent across replicates, suggesting that there is little variability beyond what is expected due to random sampling. Conversely, a high dispersion value indicates greater variability in gene expression levels among replicates.


## plot
plotBCV(y_adj)

# Fit Generalized Linear Model (GLM) 
fit <- glmFit(y_adj, design_adj)
colnames(fit)

# Differential gene expression analysis
lrt <- glmLRT(fit, contrast=c(-1,1))

## Coefficients used in the contrast for the likelihood ratio test
topTags(lrt)

## Explanation: It indicates that the comparison is between the coefficient for treatCTL (the reference group) and the coefficient for treatExe2h (the group of interest).

## Number of genes identified as differentially expressed (Disregarding parameters)
is.de <- decideTestsDGE(lrt)
summary(decideTestsDGE(lrt))

## plot MD
plotMD(lrt, status=is.de)

# Number of up and down-regulated. FDR < 5% and |log2 fold change| ≥ 0
summary(de <- decideTestsDGE(lrt, lfc=0))

# MA-plot
detags <- rownames(y_adj)[as.logical(de)]
plotSmear(lrt, de.tags=detags)
abline(h=c(-1, 1), col="blue")

# To export the results
result <- as.data.frame(topTags(lrt, n=1000000))
result <- add_column(result, symbol= rownames(result), .before = "logFC")
     
## save as table
write.table(result, "DEGs_CTL_vs_2h.txt", sep = "\t", row.names = FALSE)

## Check all the genes Downregulated and Upregulated

# Remove unnecessary columns
result$logCPM <- NULL
result$LR <- NULL

# Remove rows with FDR < 0.05 and logFC >= 0
result$expression <- ifelse(result$FDR < 0.05 & abs(result$logFC) >= 0,
                        ifelse(result$logFC >= 0, 'Up', 'Down'),
                        'Stable')

# Select only genes that are upregulated or downregulated
result_final <- subset(result, expression %in% c("Up", "Down"))

# Remove row names
rownames(result_final) <- NULL

## View the table
knitr::kable(result_final)  

##########################################
############      PLOTS        ###########
##########################################

## volcano plot

# Extract top tags from lrt
result_edgeR <- as.data.frame(topTags(lrt, n=1000000, adjust.method = "BH"))

# Extract log-transformed counts per million
count_CPM <- as.data.frame(logcpm_adj)

## Convert row names into first column
result_edgeR <- add_column(result_edgeR, gene = rownames(result_edgeR), .before = "logFC")

count_CPM <- add_column(count_CPM, gene = rownames(count_CPM), .before = "CTL_1")

## Order by gene
result_edgeR <- result_edgeR[order(result_edgeR$gene),]
count_CPM <- count_CPM[order(count_CPM$gene),]

## Combine result_edgeR columns with count_CPM
count_CPM <- add_column(count_CPM, logFC = result_edgeR$logFC, .before = "CTL_1")
count_CPM <- add_column(count_CPM, FDR = result_edgeR$FDR, .before = "CTL_1")
count_CPM <- add_column(count_CPM, gene2 = result_edgeR$gene, .before = "CTL_1")

# Check if gene and gene2 are the same and in the same order
if(all(count_CPM$gene == count_CPM$gene2)) {
  print("The values in 'gene' and 'gene2' are the same.")
} else {
  print("The values in 'gene' and 'gene2' are different.")
  different_indices <- which(count_CPM$gene != count_CPM$gene2)
  print(paste("The value in 'gene' is different at indices:", different_indices))
}

## remove gene2
count_CPM$gene2 <- NULL

## Insert -log10 FDR
count_CPM <- add_column(count_CPM, logFDR = -log10(result_edgeR$FDR), .before = "CTL_1")

## Insert -log10 p-value
count_CPM <- add_column(count_CPM, logpvalue = -log10(result_edgeR$PValue), .before = "CTL_1")


## PLOT

# Assign count_CPM to expressao_genica
expressao_genica <- count_CPM

# Determine expression based on FDR and logFC thresholds
expressao_genica$expression <- ifelse(expressao_genica$FDR < 0.05 & abs(expressao_genica$logFC) >= 0,
                                      ifelse(expressao_genica$logFC >= 0, 'Up', 'Down'),
                                      'Stable')

# Count the number of genes categorized as Up and Down
sum(expressao_genica$expression == "Up")
sum(expressao_genica$expression == "Down")

# Create volcano plot

# Filter the data to include only "Down" and "Up" genes
expressao_genica_filtered <- expressao_genica[expressao_genica$expression %in% c("Down", "Up"), ]

# Order the filtered dataframe by logFC
expressao_genica_filtered <- expressao_genica_filtered[order(expressao_genica_filtered$logFC, decreasing = TRUE), ]

# Select the top 5 upregulated genes and the top 5 downregulated genes
top_genes_up <- head(subset(expressao_genica_filtered, expression == "Up"), 5)
top_genes_down <- head(subset(expressao_genica_filtered, expression == "Down"), 5)

# Combine top_genes_up and top_genes_down into a single dataframe
top_genes <- rbind(top_genes_up, top_genes_down)

## see
print(top_genes)

# volcano plot
plot2 <- ggplot(data = expressao_genica,
       aes(x = logFC,
           y = logFDR,
           colour = expression)) +
  geom_point(size = 3) +
  geom_label_repel(data = top_genes, aes(label = gene), size = 5, force = 5, 
                   fill = "white", color = "black", box.padding = unit(0.5, "lines")) +  # Labeling top 10 genes with a rectangle
  scale_color_manual(values = c("red", "grey", "#009933"),
                     name = "",
                     breaks = c("Up", "Stable", "Down")) +
  geom_vline(xintercept = c(-0, 0), lty = 4, col = "black", lwd = 0.8) +
  labs(x = "log2FC",
       y = "-log10 (adj.P.Val)") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "right",
        legend.title = element_blank(),
        text = element_text(size = 12)) +
  ggtitle("CTL vs 2 hours")


# To display the volcano plot
plot2

## save as image
tiff("volcano_CTL_vs_2h.tiff", units="in", width=7, height=7, res=300)
plot2 
dev.off()


## heatmap

## Only genes Down and Upregulated
head(expressao_genica_filtered)

#remove the fisrt five cols
expressao_genica_filtered <- expressao_genica_filtered[, -(1:5)]

## remove a col name expression
expressao_genica_filtered$expression <- NULL

# Set column names
colnames(expressao_genica_filtered) <- c(rep("CTL", 5), rep("2h", 5)) 

# Convert dataframe to matrix
samp.with.rownames <- as.matrix(expressao_genica_filtered)

# Create column annotations for heatmap
suppressWarnings({
column_annotation <- c(rep("#74ADD1", 5), rep("purple", 5))
column_annotation <- as.matrix(column_annotation)
colnames(column_annotation) <- c("Variable X")
})

# plot
library("gplots")

heatmap.2(samp.with.rownames, scale="row", trace = "none", cexRow=1.5, cexCol=2, margins = c(5, 11), colSideColors=samp.with.rownames,ColSideColors=column_annotation, labCol = FALSE)

# Get plot coordinates
coords <- par("usr")

# Increase right margin
par(mar = c(5, 6, 4, 12) + 0.2)

# Set legend coordinates
legend_x <- coords[2] + 0.2 * diff(coords[1:2])
legend_y <- coords[4] + 0.05 * diff(coords[3:4])

# Add legend outside the plot
legend(legend_x, legend_y, legend = c("CTL", "2h"), 
       fill = c("#74ADD1", "purple"), border = FALSE, cex = 1.1, 
       xpd = TRUE)


## save as image
suppressWarnings({
tiff("heatmap_CTL_vs_2h.tiff", units="in", width=7, height=7, res=300)
heatmap.2(samp.with.rownames, scale="row", trace = "none", cexRow=1.5, cexCol=2, margins = c(5, 11), colSideColors=samp.with.rownames,ColSideColors=column_annotation, labCol = FALSE)

# Get plot coordinates
coords <- par("usr")

# Increase right margin
par(mar = c(5, 6, 4, 12) + 0.2)

# Set legend coordinates
legend_x <- coords[2] + 0.2 * diff(coords[1:2])
legend_y <- coords[4] + 0.05 * diff(coords[3:4])

# Add legend outside the plot
legend(legend_x, legend_y, legend = c("CTL", "2h"), 
       fill = c("#74ADD1", "purple"), border = FALSE, cex = 1.1, 
       xpd = TRUE)
})

```



------------------------------------------------------------------------


### **1.6 Groups (CTL vs 4 hours)**

```{r plot3, fig.cap=""}
library(tibble)
eDat <- read.table("Count.txt", header=TRUE, sep="\t", row.names = 1)
gDat <- read.table("Group.txt", header=TRUE, sep="\t")


## Counts
## filter only samples CTL and Exe0h
suppressMessages({library(dplyr)})

# Select only 'CTL' ou 'Exe4h'
selected_columns <- eDat[, grepl("^CTL|^Exe4h", names(eDat))]
  
# new table
new_table1 <- cbind(rownames = rownames(eDat), selected_columns)

## remove the first colllumn
new_table1$rownames <- NULL

## Treat
new_group1 <- subset(gDat, treat == "CTL" | treat == "Exe4h")

## load EdgeR ##
library(edgeR)

# ATENTION!!
new_group1$treat

# Control treatment needs to appear as the first level for the Treat Vs. Control comparison
new_group1$treat <-factor(new_group1$treat, levels=c("CTL", "Exe4h"))

# Input data
y_adj <- DGEList(counts=new_table1, samples=new_group1, group=new_group1$treat)
design_adj <- model.matrix(~ treat - 1, data = new_group1)

# Filter lowly expressed genes
keep_adj <- rowSums(cpm(y_adj)>1) >= 2
sum(keep_adj)
y_adj <- y_adj[keep_adj, , keep.lib.sizes=FALSE]

## Explanation: ">= 2" is a condition to maintain the gene, which means that the gene needs to have at least two replicas with CPM greater than 1 to be maintained.

## Explanation: after filtering, 16133 genes remained that have sufficient expression in at least two replicates with CPM greater than 1. Genes that do not meet these criteria are removed from the dataset.


# Normalization for RNA composition
y_adj <- calcNormFactors(y_adj)
y_adj$samples
norm.expr_adj <- y_adj$samples

# norm counts
logcpm_adj <- cpm(y_adj, log=TRUE)


##########################################
##      Data analyses with edgeR        ##
##########################################


# Estimate dispersion with Cox-Reid profile-adjusted likelihood (CR) method
y_adj <- estimateDisp(y_adj, design_adj)
y_adj$common.dispersion


## Explanation: A low dispersion value indicates that the expression levels of genes are relatively consistent across replicates, suggesting that there is little variability beyond what is expected due to random sampling. Conversely, a high dispersion value indicates greater variability in gene expression levels among replicates.


## plot
plotBCV(y_adj)

# Fit Generalized Linear Model (GLM) 
fit <- glmFit(y_adj, design_adj)
colnames(fit)

# Differential gene expression analysis
lrt <- glmLRT(fit, contrast=c(-1,1))

## Coefficients used in the contrast for the likelihood ratio test
topTags(lrt)

## Explanation: It indicates that the comparison is between the coefficient for treatCTL (the reference group) and the coefficient for treatExe0h (the group of interest).

## Number of genes identified as differentially expressed (Disregarding parameters)
is.de <- decideTestsDGE(lrt)
summary(decideTestsDGE(lrt))

## plot MD
plotMD(lrt, status=is.de)

# Number of up and down-regulated. FDR < 5% and |log2 fold change| ≥ 0
summary(de <- decideTestsDGE(lrt, lfc=0))

# MA-plot
detags <- rownames(y_adj)[as.logical(de)]
plotSmear(lrt, de.tags=detags)
abline(h=c(-1, 1), col="blue")

# To export the results
result <- as.data.frame(topTags(lrt, n=1000000))
result <- add_column(result, symbol= rownames(result), .before = "logFC")
     
## save as table
write.table(result, "DEGs_CTL_vs_4h.txt", sep = "\t", row.names = FALSE)

## Check all the genes Downregulated and Upregulated

# Remove unnecessary columns
result$logCPM <- NULL
result$LR <- NULL

# Remove rows with FDR < 0.05 and logFC >= 0
result$expression <- ifelse(result$FDR < 0.05 & abs(result$logFC) >= 0,
                        ifelse(result$logFC >= 0, 'Up', 'Down'),
                        'Stable')

# Select only genes that are upregulated or downregulated
result_final <- subset(result, expression %in% c("Up", "Down"))

# Remove row names
rownames(result_final) <- NULL

## View the table
knitr::kable(result_final)  


##########################################
############      PLOTS        ###########
##########################################

## volcano plot

# Extract top tags from lrt
result_edgeR <- as.data.frame(topTags(lrt, n=1000000, adjust.method = "BH"))

# Extract log-transformed counts per million
count_CPM <- as.data.frame(logcpm_adj)

## Convert row names into first column
result_edgeR <- add_column(result_edgeR, gene = rownames(result_edgeR), .before = "logFC")

count_CPM <- add_column(count_CPM, gene = rownames(count_CPM), .before = "CTL_1")

## Order by gene
result_edgeR <- result_edgeR[order(result_edgeR$gene),]
count_CPM <- count_CPM[order(count_CPM$gene),]

## Combine result_edgeR columns with count_CPM
count_CPM <- add_column(count_CPM, logFC = result_edgeR$logFC, .before = "CTL_1")
count_CPM <- add_column(count_CPM, FDR = result_edgeR$FDR, .before = "CTL_1")
count_CPM <- add_column(count_CPM, gene2 = result_edgeR$gene, .before = "CTL_1")

# Check if gene and gene2 are the same and in the same order
if(all(count_CPM$gene == count_CPM$gene2)) {
  print("The values in 'gene' and 'gene2' are the same.")
} else {
  print("The values in 'gene' and 'gene2' are different.")
  different_indices <- which(count_CPM$gene != count_CPM$gene2)
  print(paste("The value in 'gene' is different at indices:", different_indices))
}

## remove gene2
count_CPM$gene2 <- NULL

## Insert -log10 FDR
count_CPM <- add_column(count_CPM, logFDR = -log10(result_edgeR$FDR), .before = "CTL_1")

## Insert -log10 p-value
count_CPM <- add_column(count_CPM, logpvalue = -log10(result_edgeR$PValue), .before = "CTL_1")


## PLOT

# Assign count_CPM to expressao_genica
expressao_genica <- count_CPM

# Determine expression based on FDR and logFC thresholds
expressao_genica$expression <- ifelse(expressao_genica$FDR < 0.05 & abs(expressao_genica$logFC) >= 0,
                                      ifelse(expressao_genica$logFC >= 0, 'Up', 'Down'),
                                      'Stable')

# Count the number of genes categorized as Up and Down
sum(expressao_genica$expression == "Up")
sum(expressao_genica$expression == "Down")

# Create volcano plot

# Filter the data to include only "Down" and "Up" genes
expressao_genica_filtered <- expressao_genica[expressao_genica$expression %in% c("Down", "Up"), ]

# Order the filtered dataframe by logFC
expressao_genica_filtered <- expressao_genica_filtered[order(expressao_genica_filtered$logFC, decreasing = TRUE), ]

# Select the top 5 upregulated genes and the top 5 downregulated genes
top_genes_up <- head(subset(expressao_genica_filtered, expression == "Up"), 5)
top_genes_down <- head(subset(expressao_genica_filtered, expression == "Down"), 5)

# Combine top_genes_up and top_genes_down into a single dataframe
top_genes <- rbind(top_genes_up, top_genes_down)

# volcano plot
plot3 <- ggplot(data = expressao_genica,
       aes(x = logFC,
           y = logFDR,
           colour = expression)) +
  geom_point(size = 3) +
  geom_label_repel(data = top_genes, aes(label = gene), size = 5, force = 5, 
                   fill = "white", color = "black", box.padding = unit(0.5, "lines")) +  # Labeling top 10 genes with a rectangle
  scale_color_manual(values = c("red", "grey", "#009933"),
                     name = "",
                     breaks = c("Up", "Stable", "Down")) +
  geom_vline(xintercept = c(-0, 0), lty = 4, col = "black", lwd = 0.8) +
  labs(x = "log2FC",
       y = "-log10 (adj.P.Val)") +
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "right",
        legend.title = element_blank(),
        text = element_text(size = 12)) +
  ggtitle("CTL vs 4 hours")

# To display the volcano plot
plot3

## save as image
tiff("volcano_CTL_vs_4h.tiff", units="in", width=7, height=7, res=300)
plot3 
dev.off()


## heatmap

## Only genes Down and Upregulated
head(expressao_genica_filtered)

#remove the fisrt five cols
expressao_genica_filtered <- expressao_genica_filtered[, -(1:5)]

## remove a col name expression
expressao_genica_filtered$expression <- NULL

# Set column names
colnames(expressao_genica_filtered) <- c(rep("CTL", 5), rep("4h", 5)) 

# Convert dataframe to matrix
samp.with.rownames <- as.matrix(expressao_genica_filtered)

# Create column annotations for heatmap
suppressWarnings({
  column_annotation <- c(rep("#74ADD1", 5), rep("#fd61bf", 5))
  column_annotation <- as.matrix(column_annotation)
  colnames(column_annotation) <- c("Variable X")
})

# plot
library("gplots")

heatmap.2(samp.with.rownames, scale="row", trace = "none", cexRow=1.5, cexCol=2, margins = c(5, 11), colSideColors=samp.with.rownames,ColSideColors=column_annotation, labCol = FALSE)

# Get plot coordinates
coords <- par("usr")

# Increase right margin
par(mar = c(5, 6, 4, 12) + 0.2)

# Set legend coordinates
legend_x <- coords[2] + 0.2 * diff(coords[1:2])
legend_y <- coords[4] + 0.05 * diff(coords[3:4])

# Add legend outside the plot
legend(legend_x, legend_y, legend = c("CTL", "4h"), 
       fill = c("#74ADD1", "#fd61bf"), border = FALSE, cex = 1.1, 
       xpd = TRUE)


## save as image
suppressWarnings({
tiff("heatmap_CTL_vs_4h.tiff", units="in", width=7, height=7, res=300)
heatmap.2(samp.with.rownames, scale="row", trace = "none", cexRow=1.5, cexCol=2, margins = c(5, 11), colSideColors=samp.with.rownames,ColSideColors=column_annotation, labCol = FALSE)

# Get plot coordinates
coords <- par("usr")

# Increase right margin
par(mar = c(5, 6, 4, 12) + 0.2)

# Set legend coordinates
legend_x <- coords[2] + 0.2 * diff(coords[1:2])
legend_y <- coords[4] + 0.05 * diff(coords[3:4])

# Add legend outside the plot
legend(legend_x, legend_y, legend = c("CTL", "4h"), 
       fill = c("#74ADD1", "#fd61bf"), border = FALSE, cex = 1.1, 
       xpd = TRUE)
})

```

------------------------------------------------------------------------


### **1.7 KEGG pathway over-representation analysis using ‘clusterProfiler’**


```{r clusterprofiler, fig.cap=""}

# Load the library
library("clusterProfiler")
library("org.Mm.eg.db")
library("pathview")
library("ggplot2")
library("cowplot")


### Control vs 0 hours ###

## import DEGs
DEGs <- read.delim("~/VanessaPD/differential_expression/DEGs_CTL_vs_0h_NOFC.txt")

## select only genes with FDR <0.05 
significant_genes <- DEGs[DEGs$FDR < 0.05, ]

# For KEGG pathway enrichment using the gseKEGG() function, we need to convert id types. We can use the bitr function for this (included in clusterProfiler). It is normal for this call to produce some messages / warnings.
# In the bitr function, the param fromType should be the same as keyType from the gseGO function above (the annotation source). This param is used again in the next two steps: creating dedup_ids and df2.
# toType: in the bitr function has to be one of the available options from keyTypes(org.Dm.eg.db) and must map to one of ‘kegg’, ‘ncbi-geneid’, ‘ncib-proteinid’ or ‘uniprot’ because gseKEGG() only accepts one of these 4 options as it’s keytype parameter.

# Convert gene IDs for gseKEGG function
# We will lose some genes here because not all IDs will be converted
ids = bitr(significant_genes$symbol, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)

# Create a new dataframe df2 which has the respective entrez IDs for the gene symbols.
colnames(ids) = c("symbol", "EntrezID")
df2 = merge(significant_genes, ids, by = "symbol")

# we want the log2 fold change 
original_gene_list = df2$logFC

# name the vector
names(original_gene_list) <- df2$EntrezID

# omit any NA values 
gene_list<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)

## KEGG

kegg_ctl_vs_0h <- enrichKEGG(
  names(gene_list),
  organism = "mmu",
  keyType = "ncbi-geneid",
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.2,
  use_internal_data = FALSE
)

# table
t_kegg_ctl_vs_0h <- as.data.frame(kegg_ctl_vs_0h)

## convert gene ID to Symbol
edox <- setReadable(kegg_ctl_vs_0h, 'org.Mm.eg.db', 'ENTREZID')

# table
t_kegg_ctl_vs_0h <- as.data.frame(edox)


# Split gene IDs in t_kegg_ctl_vs_0h and create a data frame with separate rows for each gene ID
geneIDs_split <- strsplit(t_kegg_ctl_vs_0h$geneID, "/")
t_kegg_ctl_vs_0h_expanded <- data.frame(geneID = unlist(geneIDs_split), t_kegg_ctl_vs_0h[rep(seq_len(nrow(t_kegg_ctl_vs_0h)), sapply(geneIDs_split, length)), ])

# Merge the tables based on geneID
merged_data <- merge(t_kegg_ctl_vs_0h_expanded, df2, by.x = "geneID", by.y = "symbol", all.x = TRUE)

# Print the merged data
print(merged_data)

## remove (mus musculos name)
edox@result$Description <- gsub(pattern = " - Mus musculus (house mouse)", replacement = "", edox@result$Description, fixed = T)

## table
t_kegg_ctl_vs_0h <- as.data.frame(edox)


## plot
cnetplot1 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
         showCategory = c("Breast cancer", "Basal cell carcinoma", "Transcriptional misregulation in cancer", 
                          "Colorectal cancer", "Hepatocellular carcinoma", "Small cell lung cancer", 
                          "AGE-RAGE signaling pathway in diabetic complications", "Gastric cancer", 
                          "Proteoglycans in cancer", "Renal cell carcinoma", "Amphetamine addiction")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Human Diseases",
    subtitle = NULL
  )


cnetplot2 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
                      showCategory = c("FoxO signaling pathway",
                                       "TNF signaling pathway",
                                       "MAPK signaling pathway",
                                       "Hippo signaling pathway")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Environmental Information Processing",
    subtitle = NULL
  )


cnetplot3 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
                      showCategory = c("Efferocytosis",
                                       "Apoptosis",
                                       "Signaling pathways regulating pluripotency of stem cells")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Cellular Processes",
    subtitle = NULL
  )


cnetplot4 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
                      showCategory = c("Circadian rhythm",
                                       "Osteoclast differentiation",
                                       "Parathyroid hormone synthesis, secretion and action",
                                       "IL-17 signaling pathway",
                                       "Adipocytokine signaling pathway")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Organismal Systems",
    subtitle = NULL
  )


## put all of the together
KEGG_CTL_vs_0h_graph<- plot_grid(cnetplot1, cnetplot2, cnetplot3, cnetplot4, nrow = 2, ncol = 2, labels = c("A", "B", "C", "D"))


## save image
tiff('KEGG_CTL_vs_0h_graph.tiff', units="in", width=17, height=17, res=600)
KEGG_CTL_vs_0h_graph
dev.off()

## save table
write.table(t_kegg_ctl_vs_0h, file = "t_kegg_ctl_vs_0h.txt", sep = "\t", row.names = FALSE)

## dotplot
options(enrichplot.colours = c("red","blue"))

dotplot(edox, size = "Count", color = "p.adjust", showCategory = 15, title = "KEGG: Kyoto Encyclopedia of Genes and Genomes (0 hour)")

# pathview
library("pathview")
pathways <- pathview(gene.data  = gene_list,
                     pathway.id = c("mmu04657", "mmu04710"),
                     species    = "mmu",
                     limit      = list(gene=max(abs(gene_list)), cpd=1))


## Bar graph
barplot1 <- barplot(edox, showCategory=30)+
  ggtitle("KEGG: Kyoto Encyclopedia of Genes and Genomes (0 hour)")+
  theme(plot.title = element_text(hjust = 0.5, size = 15, face = "bold"))


### Control vs 2 hours ###

## import DEGs
DEGs <- read.delim("~/VanessaPD/differential_expression/DEGs_CTL_vs_2h_NOFC.txt")

## select only genes with FDR <0.05 
significant_genes <- DEGs[DEGs$FDR < 0.05, ]

# For KEGG pathway enrichment using the gseKEGG() function, we need to convert id types. We can use the bitr function for this (included in clusterProfiler). It is normal for this call to produce some messages / warnings.
# In the bitr function, the param fromType should be the same as keyType from the gseGO function above (the annotation source). This param is used again in the next two steps: creating dedup_ids and df2.
# toType: in the bitr function has to be one of the available options from keyTypes(org.Dm.eg.db) and must map to one of ‘kegg’, ‘ncbi-geneid’, ‘ncib-proteinid’ or ‘uniprot’ because gseKEGG() only accepts one of these 4 options as it’s keytype parameter.

# Convert gene IDs for gseKEGG function
# We will lose some genes here because not all IDs will be converted
ids = bitr(significant_genes$symbol, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)

# Create a new dataframe df2 which has the respective entrez IDs for the gene symbols.
colnames(ids) = c("symbol", "EntrezID")
df2 = merge(significant_genes, ids, by = "symbol")

# we want the log2 fold change 
original_gene_list = df2$logFC

# name the vector
names(original_gene_list) <- df2$EntrezID

# omit any NA values 
gene_list<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)

## KEGG

kegg_ctl_vs_2h <- enrichKEGG(
  names(gene_list),
  organism = "mmu",
  keyType = "ncbi-geneid",
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.2,
  use_internal_data = FALSE
)

# table
t_kegg_ctl_vs_2h <- as.data.frame(kegg_ctl_vs_2h)

## convert gene ID to Symbol
edox <- setReadable(kegg_ctl_vs_2h, 'org.Mm.eg.db', 'ENTREZID')

# table
t_kegg_ctl_vs_2h <- as.data.frame(edox)


# Split gene IDs in t_kegg_ctl_vs_0h and create a data frame with separate rows for each gene ID
geneIDs_split <- strsplit(t_kegg_ctl_vs_2h$geneID, "/")
t_kegg_ctl_vs_2h_expanded <- data.frame(geneID = unlist(geneIDs_split), t_kegg_ctl_vs_2h[rep(seq_len(nrow(t_kegg_ctl_vs_2h)), sapply(geneIDs_split, length)), ])

# Merge the tables based on geneID
merged_data <- merge(t_kegg_ctl_vs_2h_expanded, df2, by.x = "geneID", by.y = "symbol", all.x = TRUE)

# Print the merged data
print(merged_data)

## remove (mus musculos name)
edox@result$Description <- gsub(pattern = " - Mus musculus (house mouse)", replacement = "", edox@result$Description, fixed = T)

## table
t_kegg_ctl_vs_2h <- as.data.frame(edox)


## plot
cnetplot1 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
                      showCategory = c("Protein processing in endoplasmic reticulum",
                                       "Ribosome")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Genetic Information Processing",
    subtitle = NULL
  )


cnetplot2 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
                      showCategory = c("Coronavirus disease - COVID-19")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Human Diseases",
    subtitle = NULL
  )


## put all of the together
KEGG_CTL_vs_2h_graph<- plot_grid(cnetplot1, cnetplot2, ncol = 2, labels = c("A", "B"))


## save image
tiff('KEGG_CTL_vs_2h_graph.tiff', units="in", width=15, height=12, res=600)
KEGG_CTL_vs_2h_graph
dev.off()

## save table
write.table(t_kegg_ctl_vs_2h, file = "t_kegg_ctl_vs_2h.txt", sep = "\t", row.names = FALSE)

## dotplot
options(enrichplot.colours = c("red","blue"))

dotplot(edox, size = "Count", color = "p.adjust", showCategory = 15, title = "KEGG: Kyoto Encyclopedia of Genes and Genomes (2 hours)")

# pathview
library("pathview")
pathways <- pathview(gene.data  = gene_list,
                     pathway.id = c("mmu04657", "mmu04710"),
                     species    = "mmu",
                     limit      = list(gene=max(abs(gene_list)), cpd=1))




### Control vs 4 hours ###

## import DEGs
DEGs <- read.delim("~/VanessaPD/differential_expression/DEGs_CTL_vs_4h_NOFC.txt")

## select only genes with FDR <0.05 
significant_genes <- DEGs[DEGs$FDR < 0.05, ]

# For KEGG pathway enrichment using the gseKEGG() function, we need to convert id types. We can use the bitr function for this (included in clusterProfiler). It is normal for this call to produce some messages / warnings.
# In the bitr function, the param fromType should be the same as keyType from the gseGO function above (the annotation source). This param is used again in the next two steps: creating dedup_ids and df2.
# toType: in the bitr function has to be one of the available options from keyTypes(org.Dm.eg.db) and must map to one of ‘kegg’, ‘ncbi-geneid’, ‘ncib-proteinid’ or ‘uniprot’ because gseKEGG() only accepts one of these 4 options as it’s keytype parameter.

# Convert gene IDs for gseKEGG function
# We will lose some genes here because not all IDs will be converted
ids = bitr(significant_genes$symbol, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = org.Mm.eg.db)

# Create a new dataframe df2 which has the respective entrez IDs for the gene symbols.
colnames(ids) = c("symbol", "EntrezID")
df2 = merge(significant_genes, ids, by = "symbol")

# we want the log2 fold change 
original_gene_list = df2$logFC

# name the vector
names(original_gene_list) <- df2$EntrezID

# omit any NA values 
gene_list<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)

## KEGG

kegg_ctl_vs_4h <- enrichKEGG(
  names(gene_list),
  organism = "mmu",
  keyType = "ncbi-geneid",
  pvalueCutoff = 1,
  pAdjustMethod = "BH",
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.2,
  use_internal_data = FALSE
)

# table
t_kegg_ctl_vs_4h <- as.data.frame(kegg_ctl_vs_4h)

## convert gene ID to Symbol
edox <- setReadable(kegg_ctl_vs_4h, 'org.Mm.eg.db', 'ENTREZID')

# table
t_kegg_ctl_vs_4h <- as.data.frame(edox)


# Split gene IDs in t_kegg_ctl_vs_0h and create a data frame with separate rows for each gene ID
geneIDs_split <- strsplit(t_kegg_ctl_vs_4h$geneID, "/")
t_kegg_ctl_vs_4h_expanded <- data.frame(geneID = unlist(geneIDs_split), t_kegg_ctl_vs_4h[rep(seq_len(nrow(t_kegg_ctl_vs_4h)), sapply(geneIDs_split, length)), ])

# Merge the tables based on geneID
merged_data <- merge(t_kegg_ctl_vs_4h_expanded, df2, by.x = "geneID", by.y = "symbol", all.x = TRUE)

# Print the merged data
print(merged_data)

## remove (mus musculos name)
edox@result$Description <- gsub(pattern = " - Mus musculus (house mouse)", replacement = "", edox@result$Description, fixed = T)

## table
t_kegg_ctl_vs_4h <- as.data.frame(edox)


## plot
cnetplot1 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
                      showCategory = c("AGE-RAGE signaling pathway in diabetic complications",
                                       "Proteoglycans in cancer",
                                       "Yersinia infection")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Human Diseases",
    subtitle = NULL
  )


cnetplot2 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
                      showCategory = c("Arginine biosynthesis",
                                       "Terpenoid backbone biosynthesis",
                                       "Alanine, aspartate and glutamate metabolism")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Metabolism",
    subtitle = NULL
  )


cnetplot3 <- cnetplot(edox, color.params = list(foldChange = gene_list), circular = F, colorEdge = TRUE,
                      showCategory = c("Bile secretion",
                                       "Aldosterone synthesis and secretion",
                                       "Parathyroid hormone synthesis, secretion and action")) + 
  scale_colour_gradient2(name = "fold change", low = "darkblue", mid = "white", high = "red") +
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(size = 10, hjust = 0.5)) +
  labs(
    title = "KEGG: Kyoto Encyclopedia of Genes and Genomes\nCategory: Organismal Systems",
    subtitle = NULL
  )


## put all of the together
KEGG_CTL_vs_4h_graph<- plot_grid(cnetplot1, cnetplot2, cnetplot3, ncol = 2, nrow = 2, labels = c("A", "B", "C"))


## save image
tiff('KEGG_CTL_vs_4h_graph.tiff', units="in", width=17, height=17, res=600)
KEGG_CTL_vs_4h_graph
dev.off()

## save table
write.table(t_kegg_ctl_vs_4h, file = "t_kegg_ctl_vs_4h.txt", sep = "\t", row.names = FALSE)

## dotplot
options(enrichplot.colours = c("red","blue"))

dotplot(edox, size = "Count", color = "p.adjust", showCategory = 15, title = "KEGG: Kyoto Encyclopedia of Genes and Genomes (4 hours)")

# pathview
library("pathview")
pathways <- pathview(gene.data  = gene_list,
                     pathway.id = c("mmu04657", "mmu04710"),
                     species    = "mmu",
                     limit      = list(gene=max(abs(gene_list)), cpd=1))
```

